### ISSUE TITLE
Reduce download file size (e.g. custom output format and zip file for ~90% smaller file size)?

### ISSUE THREAD

**Issue by rcFMS** [drolbr/Overpass-API#490]
Hello,

for the new project I am working on I need data about pretty much all kinds of objects needed to draw a detailed map (roads, areas, buildings, etc.) which is why I started using the overpass API to access the OSM database. So far I've made some good progress and the result looks alright.

For the data request I've built together a URL based on examples from the wiki and some own testing. Using a WWW request with this URL I then download the data.

E.g. for the postal code 12437 (which is a part of Berlin) I'm using this:

> https://overpass-api.de/api/interpreter?data=
> [out:json][timeout:900];
> area[postal_code="12437"]->.a;
> (
> rel(area.a);
> way(area.a);
> node(area.a);
> );
> (._;>>;);
> out;

This gives me all nodes, ways and relations within the boundaries of this postal code and uses recursion to also get all ways which are part of the relations but are not within the boundary as well as all nodes of the ways.

The result looks something like this (shortened):

> {
>   "version": 0.6,
>   "generator": "Overpass API 0.7.55.3 9da5e7ae",
>   "osm3s": {
>     "timestamp_osm_base": "2018-06-08T13:01:02Z",
>     "timestamp_areas_base": "2018-06-08T11:46:01Z",
>     "copyright": "The data included in this document is from www.openstreetmap.org. The data is made available under ODbL."
>   },
>   "elements": [
> 
> {
>   "type": "node",
>   "id": 20246246,
>   "lat": 52.4740643,
>   "lon": 13.4837770
> },
> {
>   "type": "node",
>   "id": 21540767,
>   "lat": 52.4503698,
>   "lon": 13.4642317
> },
> {
>   "type": "node",
>   "id": 26953753,
>   "lat": 52.4749806,
>   "lon": 13.4737603,
>   "tags": {
>     "highway": "traffic_signals"
>   }
> },
> 
> [...]
> 
> {
>   "type": "way",
>   "id": 4067919,
>   "nodes": [
>     28933012,
>     5287364476,
>     476614441,
>     4426813657,
>     4115119029
>   ],
>   "tags": {
>     "cycleway:right": "track",
>     "highway": "primary",
>     "lanes": "3",
>     "lit": "yes",
>     "maxspeed": "50",
>     "name": "KÃ¶penicker LandstraÃŸe",
>     "oneway": "yes",
>     "postal_code": "12437",
>     "ref": "B 96a",
>     "sidewalk": "right",
>     "surface": "asphalt"
>   }
> },
> 
> [...]
> 
> {
>   "type": "relation",
>   "id": 155781,
>   "members": [
>     {
>       "type": "way",
>       "ref": 23539058,
>       "role": "inner"
>     },
>     {
>       "type": "way",
>       "ref": 42943347,
>       "role": "outer"
>     }
>   ],
>   "tags": {
>     "type": "multipolygon",
>     "waterway": "riverbank"
>   }
> },
> 
> [...]
> 
> 
>   ]
> }

After deserializing this I am able to create custom objects to hold the data and use a triangulator to create meshes from the objects which I then can display.

In order to create a scrollable map I've now started working on a tiling system which only generates certain rectangular chunks (compared to the unevenly shaped boundaries of postal code areas).

For this I've slightly changed the request URL:

> https://overpass-api.de/api/interpreter?data=
> [out:json][timeout:900];
> (
> node(50.0895,8.6407,50.1338,8.7135);
> way(50.0895,8.6407,50.1338,8.7135);
> rel(50.0895,8.6407,50.1338,8.7135);
> );
> out;

This example draws a bounding box around the city center of Frankfurt Main. It might be that I'll need to reduce the size of the bounding box since the returned data has a download size of 43MB. This is, however, actually without using recoursion, as the `(._;>>;);` line is missing! If I insert this the download size will be over 100MB! Currently this even results in the problem that the download seems to get interupted / never finish.

One thing I have noticed is the data structure in this json file:
There is a list of generic **elements** and each element has it's own type declaration (e.g. 

> "type": "node"

> "type": "way"

> "type": "relation"

WIth a total of ~44,000 occurrences in the Berlin example alone, this makes up roughly **14% of redundant data**! Instead, there could be an element "nodes" which has an array of the usual nodes but without the type declaration each time, and the same for ways and relations.

Further, there are line breaks and spaces everywhere which are nice to have for a human when looking at the json file in a normal text editor. However, these are also very redundant since a json editor will still be able to show the data in the same structure, indentations and all, when all spaces (aside from in names, descriptions, etc. of course), tabs and line breaks are removed and the json only consists of a single (albeit extremely long) line. The same goes for any json serializer/deserializer which has to actually work with this data and separates it by other symbols ("[", "]", "{", "}", ",", etc.) rather than spacers. Again in the Berlin example, this could save up to **32% of data that has to be sent by the server / downloaded by the client**!

Lastly, the json could be zipped by the server before sending it, reducing the file size by an enormous **83%**!

Combined together, this reduced the file size from 5,272,957 bytes to just 488,021 bytes. This means that there is room to reduce the file size by a **total of about 90% without losing any information**!

Instead of 100MB in the Frankfurt example with resursion, the client would only need to download 10MB and unzip the file before running the deserialization as always.

So I guess my questions are:
_Is this somehow possible already?_
_Is it possible to tell the server to use a custom output format (like the current [out:json] but without redundant types and spacers)?_
_Is it possible for the server to zip the returned file?_
_If not, would it be feasible to implement this functionality?_
_To allow for backward compatibility for others who are still using the normal json output format, could this be put behind a new parameter like e.g. [out:jsonNoSpacers] or [out:jsonNoSpacersZipped]?_

_Or am I missing something entirely here and this could be solved much much easier?_

If you've made it down here, thanks for reading and please let me know what you think.
Cheers!

**Comment by mmd-osm** [comment]
> Or am I missing something entirely here and this could be solved much much easier?

I would recommend to use use GZIP compression for HTTP responses, which makes the discussion about spaces more or less irrelevant. We also have PBF format as prototype, which would be even more efficient.

**Comment by rcFMS** [comment]
Thanks for the super fast answer!

> Your browser typically already sends this `Accept-Encoding: gzip` HTTP header to request compressed responses.

So this happens in the background and when I send a request like this

> https://overpass-api.de/api/interpreter?data=
> [out:json][timeout:900];
> area[postal_code="12437"]->.a;
> (
> rel(area.a);
> way(area.a);
> node(area.a);
> );
> (._;>>;);
> out;
the browser just unpacks it to display the text on a blank white page?

Meaning for the program I'm writing I'd just need to set the header for the WWW request and it would download a gzip file which I then need to unpack?

> Example: http://overpass-turbo.eu/s/A3D
Do I need to do something differently here instead of pressing the "Execute" button as usual?
I'm getting an error message:

> Ein Fehler ist beim Ausführen der Abfrage aufgetreten!
> Request rejected. (e.g. server not found, request blocked by browser addon, request redirected, internal server errors, etc.)
> Error-Code: error (0)

Trying to call this URL directly in the browser also gives error messages:
> https://overpass-api.de/api/interpreter?data=[out:pbf][timeout:900];(node(50.0895,8.6407,50.1338,8.7135);way(50.0895,8.6407,50.1338,8.7135);rel(50.0895,8.6407,50.1338,8.7135););(._;%3E;);out%20meta;{{data:overpass,server=http://dev.overpass-api.de/api_mmd/}}

> The data included in this document is from www.openstreetmap.org. The data is made available under ODbL.
> Error: line 1: parse error: Unknown output format: pbf
> Error: line 1: static error: Unknown output format: pbf
> Error: line 1: parse error: Unknown type "{"
> Error: line 1: parse error: An empty query is not allowed
> Error: line 1: parse error: ';' expected - '{' found.
> Error: line 1: parse error: Unexpected end of input.

**Comment by mmd-osm** [comment]
> the browser just unpacks it

Right, this is automatically done by the browser.

> Meaning for the program I'm writing I'd just need to set the header for the WWW request and it would download a gzip file which I then need to unpack?

This all done on the fly by the Apache server, see https://httpd.apache.org/docs/2.4/mod/mod_deflate.html for some hints. There's really no "file" involved here.  You just need to process the compressed response on your end, which highly depends on your programming environment.

> Do I need to do something differently here instead of pressing the "Execute" button as usual?

As PBF output is a binary forma, it doesn't make sense to display that in a browser without some pre-processing. Better use Export --> Raw data directly from Overpass API and safe the file for local processing.

> Trying to call this URL directly in the browser also gives error messages:

As I said, it's a prototype and only available on the dev instance for testing, but not on overpass-api.de. Besides, the URL is incorrect. It should read:

http://dev.overpass-api.de/api_mmd/interpreter?data=%5Bout%3Apbf%5D%5Btimeout%3A900%5D%3B%0A%28%0Anode%2850.0895%2C8.6407%2C50.1338%2C8.7135%29%3B%0Away%2850.0895%2C8.6407%2C50.1338%2C8.7135%29%3B%0Arel%2850.0895%2C8.6407%2C50.1338%2C8.7135%29%3B%0A%29%3B%0A%28._%3B%3E%3B%29%3B%0Aout%20meta%3B%0A%0A%0A%0A

Again, this is a **development server**, don't use it for deployment in any app!

**Comment by rcFMS** [comment]
> There's really no "file" involved here.

Right, I meant the byte array ;-)

> which highly depends on your programming environment.

I'm developing in C# for Unity. A quick google search gave me http://www.osmsharp.com/ which I'll probably look into next.

> Better use Export --> Raw data directly from Overpass API and save the file for local processing.

> Besides, the URL is incorrect. It should read:

Thanks for the directions and the link! 

> This is just a timeout issue. The development server runs on slow hard disks, and your bounding box is quite large. This is totally expected.

Okay, makes sense. Following the link (and allowing the cross-site-scripting that got blocked by NoScript) did indeed take a bit but at the end I got a file which I could download (which in my program would be the byte array).

Are there maybe any recommendations for an optimal chunk size? The current bounding box has ~800x800m. I was thinking maybe 500x500m? From my understanding using too small bounding boxes (when using ">>" recursion) would result in objects (ways, areas, buildings) being repeated in multiple chunks if they cross the border, right?

> Again, this is a development server, don't use it for deployment in any app!

No worries, testing during development is all I'm using it for. (I even have cache in place to send as few requests as possible.)
If the project will go live at a later point in time, we'll set up our own server with a copy of the planet.osm.

**Comment by mmd-osm** [comment]
> Are there maybe any recommendations for an optimal chunk size?

That's hard to say in general, as it highly depends on how densly an area is mapped. For sure there will be duplicates with multiple queries, which you need to filter out, if that's an issue for your post process.

**Comment by rcFMS** [comment]
(Small remark about my previous comment: I had an error when measuring the distances so the real size of the bounding box for the Frankfurt example is ~5000x5000m! No wonder this takes a while. I'll reduce the chunk size a bit.)

I've also set up a gzip decoder which is working nicely.

So yeah, the issue is solved and I'll close it.
Thanks for your help!
